Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/ComponentModeSwitcher.cpp

```diff
@@ -0,0 +1,381 @@
+            m_switcherId,
+            m_handler);
+    }
+
+    ComponentModeSwitcher::~ComponentModeSwitcher()
+    {
+        ViewportEditorModeNotificationsBus::Handler::BusDisconnect();
+        EntityCompositionNotificationBus::Handler::BusDisconnect();
+        ToolsApplicationNotificationBus::Handler::BusDisconnect();
+
+        ViewportUi::ViewportUiRequestBus::Event(
+            ViewportUi::DefaultViewportId, &ViewportUi::ViewportUiRequestBus::Events::RemoveSwitcher, m_switcherId);
+    }
+
+    void ComponentModeSwitcher::UpdateSwitcherOnEntitySelectionChange(
+        [[maybe_unused]] const EntityIdList newlySelectedEntityIds, [[maybe_unused]] const EntityIdList newlyDeselectedEntityIds)
+    {
+        AZ::Entity* entity = nullptr;
+
+        // Currently only handling one new entity selection at a time
+        if (!newlySelectedEntityIds.empty() && newlySelectedEntityIds.size() == 1)
+        {
+            AZ::ComponentApplicationBus::BroadcastResult(
+                entity, &AZ::ComponentApplicationBus::Events::FindEntity, AZ::EntityId(newlySelectedEntityIds.front()));
+
+            if (entity)
+            {
+                EntityIdList selectedEntityIds;
+                ToolsApplicationRequestBus::BroadcastResult(selectedEntityIds, &ToolsApplicationRequestBus::Events::GetSelectedEntities);
+        
+                if (selectedEntityIds.size() > 1)
+                {
+                    RemoveExclusiveComponents(*entity);
+                    return;
+                }
+
+                for (const auto& entityComponent : entity->GetComponents())
+                {
+                    // Find if the component has a component mode using GetNumOfEventHandlers
+                    const auto entityComponentIdPair = AZ::EntityComponentIdPair(entity->GetId(), entityComponent->GetId());
+                    AddComponentButton(entityComponentIdPair);
+                }
+            }
+        }
+        else if (!newlyDeselectedEntityIds.empty())
+        {
+        /*    AZStd::vector<AZ::EntityComponentIdPair> pairIdsToRemove;
+            for (auto componentDataIt : m_addedComponents)
+            {
+                pairIdsToRemove.push_back(componentDataIt.m_pairId);
+            }
+            for (auto pairId : pairIdsToRemove)
+            {
+                RemoveComponentButton(pairId);
+            }*/
+            //v.erase(std::remove_if(v.begin(), v.end(), is_odd), v.end());
+            for (AZStd::vector<ComponentData>::reverse_iterator it = m_addedComponents.rbegin(); it != m_addedComponents.rend(); ++it)
+            {
+                RemoveComponentButton(it->m_pairId);
+            }
+        }
+    }
+
+    void ComponentModeSwitcher::AddComponentButton(const AZ::EntityComponentIdPair pairId)
+    {
+        // Check if the component has a component mode.
+        const auto numHandlers = AzToolsFramework::ComponentModeFramework::ComponentModeDelegateRequestBus::GetNumOfEventHandlers(pairId);
```

Comment:
nit: This is a person preference thing but I tend to prefer using the postfix `count` rather than `num` (then you don't need the abbreviation and it sound more grammatically correct). So this would be `handlerCount` instead of the alternative `numberOfHandlers` - up to you but just wanted to mention it